### PR TITLE
SONARPY-2332 Fix FP on S5845 when comparing enum objects with types

### PR DESCRIPTION
**Pull request by thomas-serre-sonarsource**
[SONARPY-2332](https://sonarsource.atlassian.net/browse/SONARPY-2332)

Part of 

[SONARPY-2332]: https://sonarsource.atlassian.net/browse/SONARPY-2332?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: python-checks/src/test/resources/checks/tests/assertOnDissimilarTypes.py [hunk 1]
```diff
@@ -208,7 +207,7 @@ class DerivedEnumWithMembers(Enum):
 class ComparingTypeAndDerivedEnumWithMembers(unittest.TestCase):
     def test_type_of_derived_enum_and_derived_enum_class(self):
         derived_enum_member = DerivedEnumWithMembers(1)
-        self.assertIs(type(enum_member), EnumType)
+        self.assertIs(type(derived_enum_member), EnumType)
```
Comment (thomas-serre-sonarsource) [inline python-checks/src/test/resources/checks/tests/assertOnDissimilarTypes.py:210]: Mistake that was present in the previous ticket. I corrected it here.

File: python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java [hunk 2]
```diff
@@ -31,6 +31,9 @@
 
 public class RuntimeType implements InferredType {
 
+  // Name of the type returned by the type() function
+  private static final String TYPE_CLASS_NAME = "type";
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java:35]: Did you try to use `org.sonar.python.types.InferredTypes#TYPE`?
It looks like that for the v1 type inference we have this constant representing the `type` class.

File: python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java [hunk 3]
```diff
@@ -52,9 +55,27 @@ public boolean isIdentityComparableWith(InferredType other) {
...
+      boolean hasThisMetaClass = hasMetaclassInHierarchy(this);
+      return (TYPE_CLASS_NAME.equals(getTypeClass().name()) && hasOtherMetaClass)
+        || (hasThisMetaClass && TYPE_CLASS_NAME.equals(otherRuntimeType.getTypeClass().name()));
+    }
+    return false;
+  }
+
+  private static boolean hasMetaclassInHierarchy(RuntimeType runtimeType) {
+    return runtimeType.getTypeClass().hasMetaClass()
+      || runtimeType.getTypeClass().superClasses().stream().filter(ClassSymbol.class::isInstance).anyMatch(c -> ((ClassSymbol) c).hasMetaClass());
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java:76]: should we look only for current type parents for a `metaclass`, or should we go deeper in the types hierarchy?

File: python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java [hunk 4]
```diff
@@ -52,9 +55,27 @@ public boolean isIdentityComparableWith(InferredType other) {
...
+      return true;
+    }
     return this.equals(other);
   }
 
+  private boolean isComparingTypeWithMetaclass(InferredType other) {
+    if (other instanceof RuntimeType otherRuntimeType) {
+      boolean hasOtherMetaClass = hasMetaclassInHierarchy(otherRuntimeType);
+      boolean hasThisMetaClass = hasMetaclassInHierarchy(this);
+      return (TYPE_CLASS_NAME.equals(getTypeClass().name()) && hasOtherMetaClass)
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java:68]: I guess there could be just a comparison with the `org.sonar.python.types.InferredTypes#TYPE` constant.

File: python-frontend/src/test/java/org/sonar/python/types/RuntimeTypeTest.java [hunk 5]
```diff
@@ -64,6 +64,34 @@ void isIdentityComparableWith() {
     assertThat(aType.isIdentityComparableWith(new DeclaredType(b))).isTrue();
   }
 
+  @Test
+  void isIdentityComparableWithMetaclass() {
+    RuntimeType typeType = new RuntimeType(new ClassSymbolImpl("type", "type"));
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/test/java/org/sonar/python/types/RuntimeTypeTest.java:69]: here you can use `org.sonar.python.types.InferredTypes#TYPE`

File: python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java (hunk 2 above)
Comment (thomas-serre-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java:35]: good catch, I ll use it. thanks

File: python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java (hunk 3 above)
Comment (thomas-serre-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java:76]: I intend to go deeper as the original issue contained a child class from `enum.Enum`

File: python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java [hunk 6]
```diff
@@ -52,7 +52,23 @@ public boolean isIdentityComparableWith(InferredType other) {
     if (other instanceof UnionType) {
       return other.isIdentityComparableWith(this);
     }
-    return this.equals(other);
+    return isComparingTypeWithMetaclass(other) || this.equals(other);
+  }
+
+  private boolean isComparingTypeWithMetaclass(InferredType other) {
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java:58]: as discussed, could be simplified like this :
```java
private boolean isComparingTypeWithMetaclass(InferredType other) {
    if (other instanceof RuntimeType otherRuntimeType) {
      boolean hasOtherMetaClass = hasMetaclassInHierarchy(otherRuntimeType);
      boolean hasThisMetaClass = hasMetaclassInHierarchy(this);
      return (InferredTypes.TYPE.equals(this) && hasOtherMetaClass)
        || (hasThisMetaClass && InferredTypes.TYPE.equals(otherRuntimeType));
    }
    return false;
  }
```

File: python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java (hunk 3 above)
Comment (maksim-grebeniuk-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java:71]: as discussed - we need to check whole hierarchy, not nearest parents

**Review by maksim-grebeniuk-sonarsource — CHANGES_REQUESTED** [review]
After a discussion some changes requested

**Review by maksim-grebeniuk-sonarsource — APPROVED** [review]
LGTM, thanks for the fixes

**Pull request merged**